### ISSUE TITLE
debugging using dtrace doesn’t appear to be working on HBSD 12.x

### ISSUE THREAD

**Issue by AdSchellevis** [opnsense/src#91]
_since HardenedBSD seems to have removed all issues from GitHub, leaving a tracking link here (copy of https://github.com/HardenedBSD/hardenedBSD/issues/388, added again under https://git-01.md.hardenedbsd.org/HardenedBSD/HardenedBSD/issues/22)._

----

**Describe the bug**
When trying to use dtrace to monitor the kernel call stack, none of the kernel function names can be resolved by dtrace (and possibly other tools as well). Trying the same procedure on a stock FreeBSD machine does seem to function without issues. 
When debugging the kernel source, we ran into  this commit https://github.com/HardenedBSD/hardenedBSD/commit/37a8b2de291f4cb84b5cac86aab8e5edcaa75176 which describes the addition of a new privilege named `PRIV_KLD_STAT`, which, in the context of dynamically loaded kernel modules, is meant to prevent non-root users from being able to see kernel modules. This is indeed the case, except the root user get shielded from gaining insights into the machine as well due to masking the addresses in the same function (https://github.com/HardenedBSD/hardenedBSD/commit/37a8b2de291f4cb84b5cac86aab8e5edcaa75176#diff-c3d5aa900131c65102c23e944303538392a83cc73492e590a65b65f1b1505da1R1259-R1263). 

One of the fastest pointers to tell something is off seems to be to to run `kldstat` as root and witness all kernel addresses are zeroed out (0x0). This becomes problematic when tracing kernel performance using Dtrace or any other facility, as C type data (ctf(5)) does not get resolved properly, and instead gets resolved to non-descriptive kernel addresses.  
**To Reproduce**
The following steps assume the following options in GENERIC:
```
makeoptions	DEBUG=-g
makeoptions	WITH_CTF=1
options		KDTRACE_FRAME
options		KDTRACE_HOOKS
```

1.	Build a DEBUG SMP kernel with `nooptions PAX_HARDENING` and run the following (arbitrary) command:
`dtrace -n uiomove:entry’{@[stack] = count(); }’`
Run this for some time and press CTRL-C. The result will show an aggregate of unique kernel stack traces, e.g.:
```
              kernel`m_uiotombuf+0x10c
              kernel`sosend_generic+0x40d
              kernel`sosend+0x50
              kernel`kern_sendit+0x237
              kernel`sendit+0x19e
              kernel`sys_sendto+0x4d
              kernel`amd64_syscall+0x364
              kernel`0xffffffff811e38d0
                4

              kernel`ttydisc_write+0xcc
              kernel`ttydev_write+0x155
              kernel`devfs_write_f+0xda
              kernel`dofilewrite+0xb0
              kernel`sys_write+0xc1
              kernel`amd64_syscall+0x364
              kernel`0xffffffff811e38d0
                5
```

2.	Do the same as step 1, but build the kernel with `options PAX_HARDENING`. The results are as follows:
```
              0xffffffff80da98bc
              0xffffffff80db29dd
              0xffffffff80db2ed0
              0xffffffff80db9d47
              0xffffffff80dba0be
              0xffffffff80db9f0d
              0xffffffff8120b064
              0xffffffff811e3b00
                4

              0xffffffff80da20fc
              0xffffffff80d9c8b5
              0xffffffff80ba777a
              0xffffffff80d82b70
              0xffffffff80d826d1
              0xffffffff8120b064
              0xffffffff811e3b00
                5
```
**Expected behavior**
The question probably is if we really want to mask these addresses for the root user, there might be a use case for it, but if there is, it should be properly documented and perhaps being a separate build option so you don’t have to disable the PAX_HARDENING in full to stop masking addresses. If we assume that the root user itself was already properly safeguarded, question is why addresses are being masked since regular users seem not to be able to reach this point at all.
**Additional context**
The documentation at https://git-01.md.hardenedbsd.org/HardenedBSD/HardenedBSD/wiki#user-content-generic-system-hardening specifies that the PAX_HARDENING kernel option restricts what NON-root users are permitted to do. However, this option also affects the root user. 

**Environment**
OPNsense 21.1.a (amd64, OpenSSL).

**Comment by lattera** [comment]
Option 1: Perhaps, in OPNsense's src repo, we could either remove the KLD hardening patch altogether (but leaving `PRIV_KLD_STAT` defined to keep ABI/API compat with HardenedBSD.) Effectively, revert all changes except to `sys/sys/priv.h`.

Option 2: Or, we could create another kernel config knob (`HARDEN_KLD`) and use that instead of the more generic `PAX_HARDENING`. OPNsense would build without `HARDEN_KLD` while downstream derivatives (like what I'm working on at ${DAYJOB})  can make use of it. The `HARDEN_KLD` kernel config knob would exist in upstream HardenedBSD.

Using option 2 would lessen the amount of OPNsense-specific patches. I'd be more inclined to go that route.

Thoughts?

**Comment by lattera** [comment]
This is the reply I gave in the HardenedBSD ticket ( https://git-01.md.hardenedbsd.org/HardenedBSD/HardenedBSD/issues/22 ):

> If we assume that the root user itself was already properly safeguarded, question is why addresses are being masked since regular users seem not to be able to reach this point at all.

It's because we want to limit the number of kernel infoleaks. In FreeBSD, passing kernel pointers to userland is used as a feature. I'd like to get rid of all kernel infoleaks, but some utilities make direct use of them (like `gpart`, `fstat`, and `sockstat` to name a few).

We may want to enhance DTrace not to use the KLD API for resolving kernel symbols--rather, have it use a different (DTrace-specific?) API.

I believe, then, that this is a larger issue requiring a look at DTrace itself.

**Comment by lattera** [comment]
Candidate patch attached.

There's one line in this patch that isn't applicable: removing a device (device hme) from the HARDENEDBSD kernel.


[harden_kld_r01.patch.txt](https://github.com/opnsense/src/files/5683851/harden_kld_r01.patch.txt)

**Comment by fichtner** [comment]
I don't mind the priv checks so much. It's just that also clearing the pointers with the same define unconditionally that is strange. Both things should be able to be switched on and off independently. Or maybe we don't need priv checking at all if the pointers are cleared already.

**Comment by AdSchellevis** [comment]
just my two cents:

priv check alone sounds reasonable (as long as the caller can identify that he's not allowed to do this), but if the attacker already made it to root access, what's the use of trying to obfuscate the application output. I'm a bit afraid if these type of changes are in other areas as well, we might end up chasing more ghosts in the (near) future. 

In my humble opinion, If there's a good reason why data is being masked for the root user, it should be very clear why that is.

This types of strange issues have the tendency to cost people days of debugging in order to understand why something that worked like a charm in FreeBSD doesn't seem to work on our end.

**Comment by lattera** [comment]
> In my humble opinion, If there's a good reason why data is being masked for the root user, it should be very clear why that is.

Removing just one of many kernel infoleaks.

> This types of strange issues have the tendency to cost people days of debugging in order to understand why something that worked like a charm in FreeBSD doesn't seem to work on our end.

Patches gladly accepted. I need help in identifying these issues and coming up with reasonable solutions. I can't do it all alone.

**Comment by lattera** [comment]
`HARDEN_KLD` patch committed: HardenedBSD/hardenedBSD@155335ed8e909903445a31422c0edacd32f9f733

https://git-01.md.hardenedbsd.org/HardenedBSD/HardenedBSD/commit/155335ed8e909903445a31422c0edacd32f9f733

**Comment by lattera** [comment]
Follow-up commit to ungate the privilege checks: https://git-01.md.hardenedbsd.org/HardenedBSD/HardenedBSD/commit/fc99fc6c473cdcf23bb04d5020985c5dead3d697

**Comment by fichtner** [comment]
I would propose gating the privilege checks with PAX_HARDENING, but either way is fine.

**Comment by fichtner** [comment]
PS: thanks for these patches! We certainly want to use them for 21.1 👍

**Comment by lattera** [comment]
I've now cherry-picked them into the master branch of OPNsense's src.git repo.

**Comment by kevans91** [comment]
> I would propose gating the privilege checks with PAX_HARDENING, but either way is fine.

I'm a little curious- what's the benefit of gating these? The priv(9) framework is insanely flexible as one can write a MAC module (fairly trivial; see mac_ntpd as an example) to extend the privilege to whomever needs it.

**Comment by AdSchellevis** [comment]
@lattera thanks for pulling this in 👍

> > In my humble opinion, If there's a good reason why data is being masked for the root user, it should be very clear why that is.
> 
> Removing just one of many kernel infoleaks.
> 

I do understand the need to hide [certain] details from the average user, I just don't really understand the need to mask details from the root user (which is likely my lack of knowledge on the subject). In order to propose patches in these cases, one would need to have a better understanding on the intend of the feature I guess.

**Comment by fichtner** [comment]
> I'm a little curious- what's the benefit of gating these? The priv(9) framework is insanely flexible as one can write a MAC module (fairly trivial; see mac_ntpd as an example) to extend the privilege to whomever needs it.

Because if it is HBSD hardening it would be better to have a base line to test against by turning PAX_HARDENING off. It also properly annotates the change vs. vanilla FreeBSD. It's easier to grep the code than to diff against a corresponding FreeBSD commit hash and then running down the HBSD commit history. This has nothing to do with code flexibility.


Cheers,
Franco

**Comment by fichtner** [comment]
Fixed in 21.1. Thanks all!
